include/deal.II/lac/trilinos_tpetra_vector.h

```diff
@@ -0,0 +1,412 @@
+       * locally owned indices among all processors.
+       */
+      virtual size_type
+      size() const override;
+
+      /**
+       * Return the MPI communicator object in use with this object.
+       */
+      MPI_Comm
+      get_mpi_communicator() const;
+
+      /**
+       * Return an index set that describes which elements of this vector are
+       * owned by the current processor. As a consequence, the index sets
+       * returned on different processors if this is a distributed vector will
+       * form disjoint sets that add up to the complete index set. Obviously, if
+       * a vector is created on only one processor, then the result would
+       * satisfy
+       * @code
+       *  vec.locally_owned_elements() == complete_index_set(vec.size())
+       * @endcode
+       */
+      virtual ::dealii::IndexSet
+      locally_owned_elements() const override;
+
+      /**
+       * Return a const reference to the underlying Trilinos
+       * Tpetra::Vector class.
+       */
+      const Tpetra::Vector<Number> &
+      trilinos_vector() const;
+
+      /**
+       * Return a (modifyable) reference to the underlying Trilinos
+       * Tpetra::Vector class.
+       */
+      Tpetra::Vector<Number> &
+      trilinos_vector();
+
+      /**
+       * Prints the vector to the output stream @p out.
+       */
+      virtual void
+      print(std::ostream &     out,
+            const unsigned int precision  = 3,
+            const bool         scientific = true,
+            const bool         across     = true) const override;
+
+      /**
+       * Return the memory consumption of this class in bytes.
+       */
+      virtual std::size_t
+      memory_consumption() const override;
+
+      /**
+       * The vectors have different partitioning, i.e. they have use different
+       * IndexSet.
+       */
+      DeclException0(ExcDifferentParallelPartitioning);
+
+      /**
+       * Attempt to perform an operation between two incompatible vector types.
+       *
+       * @ingroup Exceptions
+       */
+      DeclException0(ExcVectorTypeNotCompatible);
+
+      /**
+       * Exception thrown by an error in Trilinos.
+       *
+       * @ingroup Exceptions
+       */
+      DeclException1(ExcTrilinosError,
+                     int,
+                     << "An error with error number " << arg1
+                     << " occurred while calling a Trilinos function");
+
+    private:
+      /**
+       * Create the CommunicationPattern for the communication between the
+       * IndexSet @p source_index_set and the current vector based
+       * on the communicator @p mpi_comm.
+       */
+      void
+      create_tpetra_comm_pattern(const IndexSet &source_index_set,
+                                 const MPI_Comm &mpi_comm);
+
+      /**
+       * Pointer to the actual Tpetra vector object.
+       */
+      std::unique_ptr<Tpetra::Vector<Number>> vector;
```

Comment:
currently you don't set types for local and global ordinals. I think we need an issue to make sure we don't forget to do something about it, I guess the current code will fail with 64bit unsigned ints.

Reply:
At least, the tests currently pass both with 32bit and 64bit indices.

Reply:
Also, not specifying local and global ordinal types will make it easier to transition to the upcoming interface.

Reply:
> not specifying local and global ordinal types will make it easier to transition to the upcoming interface.

yes, you are right.

Reply:
Tests will probably succeed because we don't have any tests that use more than 2B unknowns:-)